### PR TITLE
Spotify api

### LINKED ISSUE #12: Spotify API

### PR DESCRIPTION
**Pull request by ashley7295**
@yp0975eu @Issemoh @jy8474np 

Spotify API will still need to be adjusted as more features are added to the program but it is successfully returning the follower count of an artist. You can retrieve the Client_ID and Client Secret from [https://developer.spotify.com/dashboard/](url) and by creating an account.

### CODE REVIEW

File: spotifyAPI.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    def get_client_credentials(self):
+        #returns a base64 encoded string for the auth header
+        client_id = self.client_id
+        client_secret = self.client_secret
+        if client_secret == None or client_id == None:
+            raise Exception ("You must set a client_id and client_secret") #these are required params
+        client_credentials = f'{client_id}:{client_secret}'
+        client_credentials_b64 = base64.b64encode(client_credentials.encode())
+        return client_credentials_b64.decode()
+
```
Comment (yp0975eu) [inline spotifyAPI.py]: Line 40 - 46 only needs to happen once on init.
Init could set the b64 encoded string as a variable when initialized and get_client_credentials can return the variable

File: spotifyAPI.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        if expires < now:
+            self.preform_authorization()
+            return self.get_access_token()
+        elif token == None:
+            self.preform_authorization()
+            return self.get_access_token()
+        return token
+
+    #temp function, this may be moved into ui.py later
+    def get_user_search_q(self):
```
Comment (yp0975eu) [inline spotifyAPI.py]: SpotifyAPI class should not be concerned with I/O operations. I would remove this method altogether.

File: spotifyAPI.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        return artist.lower()
+
+    #GET requests for searching for things from the API. requires Auth.
+    def search_artist_data(self, querie):
+        querie_type = 'artist' #querie type will always be artist for purposes of this program
+        access_token = self.get_access_token()
+
+        headers = {'Authorization': f'Bearer {access_token}'}
+        endpoint = 'https://api.spotify.com/v1/search'
+        data = urlencode({'q':querie, 'type':querie_type}) #using urlencode to make it a url ready string
```
Comment (yp0975eu) [inline spotifyAPI.py]: requests module will encode url parameters for you. There's no need to encode them here.

File: spotifyAPI.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+        lookup_url = f'{endpoint}?{data}'
```
Comment (yp0975eu) [inline spotifyAPI.py]: Requests takes a url, and params named argument.  On this line you re constructing a url with query parameter data. Requests module will do this for you.

```
endpoint = https://api.spotify.com/v1/search
params = {'q':querie, 'type':querie_type}
requests.get(endpoint, params=params)
```

**Comment by yp0975eu** [comment]
@ashley7295 , you might needs to merge master into your branch, then re-push so that you can resolve any conflicts.

**Review by yp0975eu — CHANGES_REQUESTED** [review]
I've added a few suggestions.

File: spotifyAPI.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            self.preform_authorization()
+            return self.get_access_token()
+        elif token == None:
+            self.preform_authorization()
+            return self.get_access_token()
+        return token
+
+
+    #GET requests for searching for things from the API. requires Auth.
+    def search_artist_data(self, querie):
```
Comment (yp0975eu) [inline spotifyAPI.py]: I think `querie` is spelled with a `y`.  `querie`  -> `query`

**Review by yp0975eu — CHANGES_REQUESTED** [review]
spelling mistake

File: spotifyAPI.py (hunk 1 above)
Comment (ashley7295) [reply spotifyAPI.py]: Issue here is resolved as your suggestion works with python 2.0 but not with python 3.8. Program uses Python 3.8.2.

File: spotifyAPI.py (hunk 3 above)
Comment (ashley7295) [reply spotifyAPI.py]: resolved issue with in person code review. 'data' value will be encoded in requests. 

       data = {'q':query,'type':querie_type}
        lookup_url = f'{endpoint}'
        r = requests.get(lookup_url,headers=headers, params = data)
        followers = r.json()

**Comment by ashley7295** [comment]
Adjusted all code review, leaving open for further review as per request

**Review by yp0975eu — APPROVED** [review]

**Comment by jy8474np** [comment]
Zoom code review. In progress code looks good. Still needs testing and validation.

**Pull request merged**